File: sols/DMOJ/dmopc21c7/P3_Whack_an_Ishigami.cpp
/* DMOPC ' 21 Contest 7 P3 - Whack-an-Ishigami
https://dmoj.ca/problem/dmopc21c7p3
*/

// g++ code.cpp
#pragma GCC optimize("O3")
#pragma GCC target("avx2")

#include <bits/stdc++.h>
using namespace std;

#define FOR(i, a, b) for (int i = (a), _##i = (b); i < _##i; ++i)
#define FORD(i, a, b) for (int i = (a), _##i = (b); i >= _##i; --i)
#define REP(i, a) for (int i = 0, _##i = (a); i < _##i; ++i)
#define REPD(i, n) for (int i = (n)-1; i >= 0; --i)
#define all(x) (x).begin(), (x).end()
#define pb push_back

typedef long long LL;

int N, M;
void solve() {
    cin >> N >> M;

    vector<int> S(N);
    REP(i, N) {
        cin >> S[i];
    }

    vector<vector<int>> adj(N, vector<int>());
    vector<vector<int>> rev_adj(N, vector<int>());
    vector<int> deg(N);
    int a, b;
    REP(i, M) {
        cin >> a >> b;
        --a, --b;
        adj[a].pb(b);
        rev_adj[b].pb(a);
        deg[a]++;
    }

    // toposort on reverse adj (i.e. pop nodes leading to nothing)
    // final order contains all nodes not leading into or inside cycle
    vector<int> topo;
    stack<int> st;
    REP(u, N) {
        if (!deg[u]) {
            st.push(u);
        }
    }

    while (!st.empty()) {
        int u = st.top();
        st.pop();
        topo.pb(u);
        for (int v : rev_adj[u]) {
            deg[v]--;
            if (!deg[v]) {
                st.push(v);
            }
        }
    }

    // push down flips using topo order
    reverse(all(topo));
    vector<int> flip(N);
    int res = 0;
    for (int u : topo) {
        if ((S[u] + flip[u]) & 1) {
            res++;
            flip[u]++;
        }
        for (int v : adj[u]) {
            flip[v] += flip[u];
        }
    }

    // now only all nodes not leading into or inside cycle handled
    // remaining nodes cannot be flipped
    // so check that orig state is 0
    REP(u, N) {
        if ((S[u] + flip[u]) & 1) {
            cout << -1 << endl;
            return;
        }
    }

    cout << res << endl;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(NULL);
    solve();
    return 0;
}

File: sols/DMOJ/dmopc21c7/P3_Whack_an_Ishigami.py
''' DMOPC '21 Contest 7 P3 - Whack-an-Ishigami
https://dmoj.ca/problem/dmopc21c7p3
'''

############ MLE ############

import sys
input = sys.stdin.readline

INF = float('inf')

def main():
    N, M = map(int, input().split())
    S = list(map(int, input().split()))

    adj = [[] for _ in range(N)]
    rev_adj = [[] for _ in range(N)]
    deg = [0] * N
    for _ in range(M):
        a, b = map(int, input().split())
        adj[a-1].append(b-1)
        rev_adj[b-1].append(a-1)
        deg[a-1] += 1

    # toposort on reverse adj (i.e. pop nodes leading to nothing)
    # final order contains all nodes not leading into or inside cycle
    topo = []
    st = [u for u in range(N) if deg[u] == 0]
    while st:
        u = st.pop()
        topo.append(u)
        for v in rev_adj[u]:
            deg[v] -= 1
            if deg[v] == 0: st.append(v)
        
    del rev_adj, st, deg
    
    # push down flips using topo order
    topo.reverse()
    flip = [0] * N
    res = 0
    for u in topo:
        if (S[u] + flip[u]) % 2 == 1:
            res += 1
            flip[u] += 1
        for v in adj[u]:
            flip[v] += flip[u]
    
    # now only all nodes not leading into or inside cycle handled
    # remaining nodes cannot be flipped
    # so check that orig state is 0
    for u in range(N):
        if (S[u] + flip[u]) % 2 == 1:
            return -1

    return res


if __name__ == '__main__':
    out = main()
    print(out)